### ISSUE TITLE
Computed properties: only on specific route.query properties

### ISSUE THREAD

**Issue by franciscolourenco** [vuejs/vue-router#562]
Computed properties get recalculated when different properties of `route.query` get changed, even if the computed property doesn't depend on those properties.

For example, `page: () => this.$route.query.page || 1` gets recalculated if we set `$route.query.random = 'something'`

http://jsfiddle.net/0brxezmj/73/

**Comment by fnlctrl** [comment]
This is expected since the whole `$route` object is replaced every time the router redirects. So every computed property depending on `$route` is notified.
http://jsfiddle.net/fnlCtrl/wshweaax/

Just like when you use computed on any other property.
http://jsfiddle.net/fnlCtrl/rc4am086/

**Comment by franciscolourenco** [comment]
> This is expected since the whole $route object is replaced every time the router redirects. So every computed property depending on $route is notified.
> http://jsfiddle.net/fnlCtrl/wshweaax/

That is the behaviour which this issue intends to see improved. Why did you close it?

> Just like when you use computed on any other property.
> http://jsfiddle.net/fnlCtrl/rc4am086/

What were you trying to achieve in this example? Its broken in a couple of ways.

Computed properties are only recalculated when the properties they depend on are changed. If the whole object is changed, the computed properties get recalculated even though they don't need to. That is why the whole $route object shouldn't be replaced, just the parts which get changed. This is an implementation detail which can have big performance penalties if the computed properties are expensive. Therefore it should probably be improved sooner or later.

**Comment by fnlctrl** [comment]
The is related to vue's implementation of it's reactivity. The second example was to show that if an object gets replaced, computed properties that depend on it gets notified. Sorry to have not made that clear.

As for vue-router, Using new $route every time won't cause noticeable performance overhead (even when there is a lot of data under `params` and `query`). It's just acting `immutable` so that you can differentiate `transition.to` and `transition.from`.

Simply put, if just by using a new object every time causes performance issues, `react` and `immutablejs` wouldn't have been invented.

**Comment by franciscolourenco** [comment]
The perfomance penalties don't come directly from using a different object, or immutability, but from the unnecessary recalculation of cached computed properties which might depend on them. You can also call it loss of efficiency if you prefer. If React doesn't have cached computed properties in the first place, there is no efficiency to lose.

If you want to achieve the same performance benefits with react, you have to check for differences manually and decide which computed properties to update.

Vue has a lot of qualities, and a mutable data model is one of its defining characteristics. Lets not try to turn Vue into React?

> It's just acting immutable so that you can differentiate transition.to and transition.from.

Without getting too much into this since I'm not familiar with the source, that seems like an implementation detail which maybe could be changed for example to cloning a snapshot of the route object into `transition.from` before it is mutated?

**Comment by fnlctrl** [comment]
I'm just stating that using a new object every time doesn't always cause performance overheads, that's all and nothing more.

Maybe using the word `acting immutable` is not entirely correct, so I'm really sorry about that and please let me rephrase it.
It's not about cloning and stuff. The router is just using your config (a new object) passed in router.go() to create another `$route`. The old one and the new one was passed to router hooks, and then the new one replaces the old one's reference inside router, though you can access both of their references inside router hooks because they are passed in as parameter. When you are done using the hooks, the old one just got GC-ed because there is no more reference.

**Comment by franciscolourenco** [comment]
I understand what is happening. Just saying that what is happening is not cooooool because of computed properties which might depend on query parameters which don't change and wouldn't have to be recalculated otherwise. The route implementation doesn't align with the way Vue computed properties are designed. The implementation should probably be rethought.

**Comment by fnlctrl** [comment]
The `$route` object needs to be replaced since we need both `transition.from` and `transition.to`, so unfortunately there is no way around.

**Comment by franciscolourenco** [comment]
Like I said, why don't we create `transition.from` by cloning `$route`? It's almost just doing the opposite of what we are doing now.

**Comment by fnlctrl** [comment]
It's actually not a performance-critical point in my opinion, so if you believe it's really significant, you can provide performance tests to convince the developers to re-implement it.
But pull requests are always welcome if you want implement it yourself. :smile:

**Comment by yyx990803** [comment]
Any optimization comes with a cost, in this case it may be a more complex internal implementation, or a less fluent external API. Either way, I simply don't see the potential perf implications being significant enough to warrant the effort needed to avoid it.

**Comment by franciscolourenco** [comment]
Ok, imagine you have a huge number of items. You are computing some sums based on all items, sorted version of the items and filtered version of the items, and the slice for pagination, all this must be configurable through the url. Now lets say the user changes the page. The only thing which needs to be re-calculated is a new slice, which is quite fast. However with current implementation, everything will be recalculated, and switching pages is no longer instant but can take a while depending on the number of items.

**Comment by yyx990803** [comment]
Why would "sums of all items" and "sorted version of all items" be dependent on `$route`? You can simply make them separate computed properties (and thus cached).

**Comment by franciscolourenco** [comment]
Well, if its sum of total items it doesn't have to be recalculated, but imagine that you want the sum of the filtered items, and the filtered items get recalculated because of url change. **page -> url -> re-filter -> re-sum.**

Sorted version would be recalculated because would be sorted according the column and order specified in the url. To prevent this, we could create a separate computed property for each field + order sorting combination, but that sounds like a workaround and way less usable than just having a computed property which depends on sorting and order properties.

**Comment by yyx990803** [comment]
Does the filter depend on `$route`? If it does then everything needs to be re-calculated anyway, if it doesn't then the filter can be a separate computed property.

You can come up with all kind of hypothetical scenarios, but this is really entering premature optimization territory. I'm happy to improve it when you have a real test case that shows immutable `$route` is the root of a real performance issue, but before then this just seems overthinking it.

**Comment by franciscolourenco** [comment]
Yes, filter depends on a query parameter from the route. Nothing would need to be recalculated if the filter didn't change, but currently everything needs to be recalculated even if the filter doesn't change.

Its not an hypothetical scenario. I wouldn't have opened an issue if it wasn't a real case. I can come up with all kinds of workarounds too, but maybe there will be users (developers) with slow performance without having a clue why, when they could have better performance by default. In the end, the burden of implementation shifts from the library to the developers. 

I see I will need to make a demo later though...
